### ISSUE TITLE
Support for dual-stack services in the speaker

### ISSUE THREAD

**Issue by uablrek** [metallb/metallb#731]
In K8s v1.20.0 it will be possible to define dual-stack services (see also https://github.com/metallb/metallb/issues/698);
```
apiVersion: v1
kind: Service
metadata:
  name: svc-preferdual-lb
spec:
  ipFamilyPolicy: PreferDualStack
  selector:
    app: mconnect-daemonset
  ports:
  - port: 5001
  type: LoadBalancer
```
The `controller` will assign two load-balancer addresses (from the same pool) when https://github.com/metallb/metallb/pull/727 is merged;
```
$ kubectl get svc svc-preferdual-lb -o json | jq .status
{
  "loadBalancer": {
    "ingress": [
      {
        "ip": "10.0.0.2"
      },
      {
        "ip": "1000::1"
      }
    ]
  }
}
```
Currently the speaker only handles one address and must be extended to handle two.

### Implementation

Implementation can start right away! No need to wait for K8s v1.20.0 or https://github.com/metallb/metallb/pull/727.

Ipv6 is supported for L2 only so BGP must wait for now. Most updates are probably in code common for L2 and BGP.

### Backward compatibility

Single-stack addresses are handles like before and there is no API update for `status.loadBalancer.ingress` so backward compatibility should not be a problem.

### Testing

E2e testing requires a dual-stack environment. The function is however not super complex so perhaps manual testing is sufficient from start.

**Comment by russellb** [comment]
Would you like to file a BGP issue, as well?  It does make sense to track it separately, but hopefully we're not too far off from unblocking it.  It should just depend on #590

**Comment by uablrek** [comment]
Can I instead remove (L2) from the title? I think the necessary updates will be in code shared for both L2 and BGP.

**Comment by russellb** [comment]
> Can I instead remove (L2) from the title? I think the necessary updates will be in code shared for both L2 and BGP.

Sure.

**Comment by tzakrajs** [comment]
At least one more change is needed to setBalancer function in speaker/bgp_controller.go is not IPv6 aware `m := net.CIDRMask(adCfg.AggregationLength, 32)` this will definitely explode when the aggregation-length is greater than 32 (e.g. 64 which is a typical prefix for IPv6 endpoints) net.CIDRMask supports IPv6 so we need a conditional around this to set a maxAggregationLength of 128 for IPv6 pools. See: https://github.com/metallb/metallb/blob/ef49019fb067ba4bf8110c53a25f45ef8c934fa0/speaker/bgp_controller.go#L220

**Comment by SchoolGuy** [comment]
@fedepaol / @johananl / @rata 

Can't this issue be closed because https://github.com/metallb/metallb/pull/1065 was merged? Or is there something missing for dual-stack support in k8s services in the speaker context?

**Comment by PKizzle** [comment]
Will the changes of #1065 be available in v0.11.1 or v0.12.0?

**Comment by PKizzle** [comment]
What is still missing is an annotation to allow to specify a custom loadBalancerIP for IPV4 and IPV6 as kubernetes has dropped support for this field (https://github.com/kubernetes/kubernetes/pull/107235)

**Comment by fedepaol** [comment]
> What is still missing is an annotation to allow to specify a custom loadBalancerIP for IPV4 and IPV6 as kubernetes has dropped support for this field ([kubernetes/kubernetes#107235](https://github.com/kubernetes/kubernetes/pull/107235))

Thanks for raising that! It looks like adding a custom annotation is the officially proposed way to do dual stack loadBalancerIP.
I'll file an issue for that.

**Comment by fedepaol** [comment]
> Will the changes of #1065 be available in v0.11.1 or v0.12.0? Currently I am still receiving a `no available IPs` error when trying to use `PreferDualStack`

We should cut out a release pretty soon, not sure what version it will be.

**Comment by fedepaol** [comment]
Closing this, filed https://github.com/metallb/metallb/issues/1155 for the dual stack / annotation gap.
